climada_petals/engine/supplychain.py

```diff
@@ -250,101 +503,144 @@ def calc_indirect_impact(self, io_approach='ghosh'):
         Analysis, Resources, 2, 489-503; doi:10.3390/resources2040489, 2013.
         """
 
-        io_switch = {'leontief': self._leontief_calc, 'ghosh': self._ghosh_calc,
-                     'eeioa': self._eeioa_calc}
-
-        # Compute coefficients based on selected IO approach
-        coefficients = np.zeros_like(self.mriot_data, dtype=np.float32)
-        if io_approach in ['leontief', 'eeioa']:
-            for col_i, col in enumerate(self.mriot_data.T):
-                if self.total_prod[col_i] > 0:
-                    coefficients[:, col_i] = np.divide(col, self.total_prod[col_i])
-                else:
-                    coefficients[:, col_i] = 0
+        self.calc_matrixes(io_approach=io_approach)
+
+        # find a better place to locate conv_fac, once and for all cases
+        if io_approach == "leontief":
+            degr_demand = (
+                self.secs_stock_shock * self.mriot.Y.values.flatten() * self.conv_fac()
+            )
+
+            self.indir_prod_impt_mat = pd.concat(
+                [
+                    pymrio.calc_x_from_L(self.inverse, degr_demand.iloc[i])
+                    for i in range(len(impact.event_id))
+                ],
+                axis=1,
+            ).T.set_index(impact.event_id)
+
+        elif io_approach == "ghosh":
+            value_added = calc_v(self.mriot.Z, self.mriot.x)
+            degr_value_added = (
+                self.secs_stock_shock * value_added.values * self.conv_fac()
+            )
+
+            self.indir_prod_impt_mat = pd.concat(
+                [
+                    calc_x_from_G(self.inverse, degr_value_added.iloc[i])
+                    for i in range(len(impact.event_id))
+                ],
+                axis=1,
+            ).T.set_index(impact.event_id)
+
+        elif io_approach == "eeioa":
+            self.indir_prod_impt_mat = (
+                pd.DataFrame(
+                    self.secs_stock_shock.dot(self.inverse)
+                    * self.mriot.x.values.flatten()
+                )
+                * self.conv_fac()
+            )
+
```

Comment:
Need to handle unknown approach?

```suggestion
        else:
            raise RuntimeError(f"Unknown io_approach: {io_approach}")

```